### PR TITLE
Add `ALL` ruff rules

### PR DESCRIPTION
**Pull request by paddyroddy**
Not sure when `ALL` got added, but certainly better than keeping up-to-date. Have added and fixed the project accordingly.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,14 +1,13 @@
 ci:
   autofix_prs: false
-
+exclude: "{{ cookiecutter.project_slug }}"
```
Comment (paddyroddy) [inline .pre-commit-config.yaml:3]: This is probably better than specifying everywhere. We have the linting of the template CI anyway.

File: pyproject.toml [hunk 2]
```diff
@@ -15,51 +15,13 @@ ignore = [
     "D417", # argument description in docstring (unreliable)
     "ISC001", # simplify implicit str concatenation (ruff-format recommended)
 ]
-per-file-ignores = {"post_gen_project.py" = [
-    "T201",
-], "test_*" = [
+per-file-ignores = {"*tests*" = [
```
Comment (paddyroddy) [inline pyproject.toml:18]: Handling both the repo & the template

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -46,7 +46,3 @@ repos:
         args:
           - --fix=lf
       - id: trailing-whitespace
-  - repo: https://github.com/renovatebot/pre-commit-hooks
-    rev: 37.36.2
-    hooks:
-      - id: renovate-config-validator
```
Comment (paddyroddy) [inline .pre-commit-config.yaml]: Unnecessary for this repo and updates too frequently

File: pyproject.toml [hunk 4]
```diff
@@ -15,51 +15,13 @@ ignore = [
...
-    "RUF",
-    "S",
-    "SIM",
-    "T",
-    "TCH",
-    "TID",
-    "UP",
-    "W",
-    "YTT",
+    "ALL",
```
Comment (samcunliffe) [inline pyproject.toml:24]: ✨

File: tests/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+"""Test the generation of the cookiecutter template."""
```
Comment (samcunliffe) [inline tests/__init__.py]: Why needed?

File: {{cookiecutter.project_slug}}/tests/__init__.py [hunk 6]
```diff
@@ -0,0 +1 @@
+"""Test the {{ cookiecutter.package_name }} package."""
```
Comment (samcunliffe) [inline {{cookiecutter.project_slug}}/tests/__init__.py]: Ditto.

**Review by samcunliffe — APPROVED** [review]
Not sure why we need `tests/__init__.py` all of a sudden(?) 

I'd do one clean PR per thing and try to keep the diff clean. So one PR replacing the list in the pyproject.toml and then one making changes to the test functions.

File: {{cookiecutter.project_slug}}/tests/__init__.py (hunk 6 above)
Comment (paddyroddy) [reply {{cookiecutter.project_slug}}/tests/__init__.py]: https://docs.astral.sh/ruff/rules/undocumented-public-package

File: tests/__init__.py (hunk 5 above)
Comment (paddyroddy) [reply tests/__init__.py]: https://docs.astral.sh/ruff/rules/undocumented-public-package

**Comment by paddyroddy** [comment]
> Not sure why we need `tests/__init__.py` all of a sudden(?)

https://docs.astral.sh/ruff/rules/implicit-namespace-package

**Comment by paddyroddy** [comment]
> I'd do one clean PR per thing and try to keep the diff clean. So one PR replacing the list in the pyproject.toml and then one making changes to the test functions.

Okay, but that's going to be a fair few PRs - so please do review them!

**Comment by paddyroddy** [comment]
Currently, breaks [INP001](https://docs.astral.sh/ruff/rules/implicit-namespace-package). This is intended, merging.

**Comment by paddyroddy** [comment]
> Okay, but that's going to be a fair few PRs - so please do review them!

Please see #219, #220 @samcunliffe

File: tests/__init__.py (hunk 5 above)
Comment (samcunliffe) [reply tests/__init__.py]: Right but our tests is not a public package.

Comment (paddyroddy) [reply tests/__init__.py]: The fact that they aren't `_tests` means that they are?

Comment (paddyroddy) [reply tests/__init__.py]: @matt-graham you know this stuff better than me

Comment (matt-graham) [reply tests/__init__.py]: I'd agree with Sam that we ideally don't want a `__init__.py` in `tests` as my feeling is that we don't want it to be considered a package, as we'd not usually want to import test modules directly, but as `tests` will still be importable as a namespace package irrespective does make this all a bit ambiguous. [Perhaps more usefully, Pytest's suggested directory structure when having tests separate from main package doesn't include an `__init__.py`](https://docs.pytest.org/en/stable/explanation/goodpractices.html#tests-outside-application-code) so probably good to follow that? We possibly also should then follow the recommendation to add `"--import-mode=importlib"` to `tool.pytest.ini_options.addopts` in `pyproject.toml` to [avoid issues with test modules with non-unique names](https://docs.pytest.org/en/stable/explanation/goodpractices.html#which-import-mode).

Comment (matt-graham) [reply tests/__init__.py]: I missed that this is now closed and conversation is going on elsewhere so ignore above.

Comment (samcunliffe) [reply tests/__init__.py]: My bad. Poor thread discipline, and poor redirection.
→ https://github.com/UCL-ARC/python-tooling/pull/219#issuecomment-1794687931 ff

Comment (paddyroddy) [reply tests/__init__.py]: I'm going to add an issue for @matt-graham's suggestion!

**Pull request merged**